### PR TITLE
Php1branch > tl1revbranch

### PR DESCRIPTION
**Pull request by php-1**

### CODE REVIEW

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+/**
+ * Author:  php1
+ * Created: Jan 21, 2018
+ */
+
+CREATE TABLE llx_ias_kpis(
```
Comment (hythamsoliman81) [inline portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: مفيش جمع في اسامي الجداول
مينفعشي تسمي kpis
لازم تبقي kpi

Comment (hythamsoliman81) [inline portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: don't use ias
it should be "prefix"_hcm_kpi

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	rowid INTEGER(11) AUTO_INCREMENT PRIMARY KEY,
+	-- BEGIN MODULEBUILDER FIELDS
+	fk_user INTEGER(11) NOT NULL,
```
Comment (hythamsoliman81) [inline portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: You don't have to mention that is a fk while you naming the field
fk_user should be user_id

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_iaskpis_assign.sql [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+/**
+ * Author:  php1
+ * Created: Jan 21, 2018
+ */
+
+
```
Comment (hythamsoliman81) [inline portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_iaskpis_assign.sql]: CREATE TABLE 'prefix'_hcm_kpi_assign (
rowid INTEGER(11) NOT NULL AUTO_INCREMENT PRIMARY KEY,
userid INTEGER(11) NOT NULL,
creationdate DATETIME NOT NULL,
effectivedate DATETIME NOT NULL,
status CHAR (1),
user_id INTEGER(11) NOT NULL,
kpi_id INTEGER(11) NOT NULL,
desc VARCHAR(255),
fromdt DATETIME NOT NULL,
todt DATETIME NOT NULL
)ENGINE=innodb;

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql [hunk 4]
```diff
@@ -0,0 +1,22 @@
+/* 
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+/**
+ * Author:  php1
+ * Created: Jan 21, 2018
```
Comment (hythamsoliman81) [inline portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: CREATE TABLE 'prefix'_hcm_kpi(
rowid INTEGER(11) NOT NULL AUTO_INCREMENT PRIMARY KEY,
userid INTEGER(11) NOT NULL,
creationdate DATETIME NOT NULL,
effectivedate DATETIME NOT NULL,
status CHAR (1) NOT NULL,
name VARCHAR(50) NOT NULL,
desc VARCHAR(255),
weight INTEGER(1) NOT NULL,
fromdt DATETIME NOT NULL,
todt DATETIME NOT NULL
) ENGINE=innodb;

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php [hunk 5]
```diff
@@ -83,11 +76,7 @@
 if (empty($action) && empty($id) && empty($ref)) $action='view';
 
 // Protection if external user
-if ($user->societe_id > 0)
-{
-	//accessforbidden();
-}
-//$result = restrictedArea($user, 'ihcm', $id);
+if ($user->societe_id > 0) {  accessforbidden(); }
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php]: I will remove this code today

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php [hunk 6]
```diff
@@ -110,715 +99,458 @@
...
 
-	print '<div class="center"><input type="submit" class="button" name="save" value="'.$langs->trans("Save").'">';
-	print ' &nbsp; <input type="submit" class="button" name="cancel" value="'.$langs->trans("Cancel").'">';
-	print '</div>';
+                $iasobject->fetchAssignKpiName($id);
+                $canedit=(($user->id == $iasobject->fk_user && $user->rights->ihcm->write->kpi) || ($user->id != $iasobject->fk_user && $user->rights->ihcm->write->kpi));
 
-	print '</form>';
-}
+                $userRequest = new User($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php]: to fetch  //user name  from line 424 to line 430

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/class/iaskpis.class.php [hunk 7]
```diff
@@ -26,6 +26,7 @@ class iaskpis extends CommonObject{
     var $fk_user;
     var $kpi_name;
     var $kpi_weight;
+    var $date_creation;
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/class/iaskpis.class.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/class/iaskpis.class.php [hunk 8]
```diff
@@ -55,16 +56,18 @@ function create($user , $notrigger=0)
...
         if (empty($this->fk_user) || ! is_numeric($this->fk_user) || $this->fk_user < 0) { $this->error="ErrorBadParameterFkUser"; return -1; }
         $this->db->begin();
-           
+        $this->date_creation=date("Y-m-d H:i:s");
         // Insert request
         $sql = "INSERT INTO ".MAIN_DB_PREFIX."ias_kpis(";
         $sql.= "fk_user,";
         $sql.= "kpi_name,";
-        $sql.= "kpi_weight";
+        $sql.= "kpi_weight,";
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/class/iaskpis.class.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/class/iaskpis.class.php [hunk 9]
```diff
@@ -502,13 +505,16 @@ function assignkpiname($user , $notrigger=0)
         // Check parameters
         if (empty($this->fk_user) || ! is_numeric($this->fk_user) || $this->fk_user < 0) { $this->error="ErrorBadParameterFkUser"; return -1; }
         $this->db->begin();
+        $this->date_creation=date("Y-m-d H:i:s");
         // Insert request
         $sql = "INSERT INTO ".MAIN_DB_PREFIX."iaskpis_assign(";
         $sql.= "fk_user,";
-        $sql.= "fk_kpi";
+        $sql.= "fk_kpi,";
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/class/iaskpis.class.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php [hunk 10]
```diff
@@ -54,14 +40,21 @@
 include_once(DOL_DOCUMENT_ROOT.'/core/class/html.formcompany.class.php');
 dol_include_once('/ihcm/class/myobject.class.php');
 
+include_once DOL_DOCUMENT_ROOT.'/custom/ihcm/kpi/class/iaskpis.class.php';
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php [hunk 11]
```diff
@@ -54,14 +40,21 @@
...
 // Load traductions files requiredby by page
 $langs->loadLangs(array("ihcm","other"));
+$langs->load('users');
 
 // Get parameters
-$id			= GETPOST('id', 'int');
-$action		= GETPOST('action', 'alpha');
-$cancel     = GETPOST('cancel', 'aZ09');
-$backtopage = GETPOST('backtopage', 'alpha');
+$iasobject = new iaskpis($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php [hunk 12]
```diff
@@ -110,715 +99,458 @@
...
-		}
-		else
-		{
-			if (! empty($object->errors)) setEventMessages(null, $object->errors, 'errors');
-			else setEventMessages($object->error, null, 'errors');
-		}
-	}
-}
-
+            $iasobject = new iaskpis($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php [hunk 13]
```diff
@@ -110,715 +99,458 @@
...
+ *
+ * Put here all code to build page
+ */
 
-	print '<form method="POST" action="'.$_SERVER["PHP_SELF"].'">';
-	print '<input type="hidden" name="action" value="update">';
-	print '<input type="hidden" name="backtopage" value="'.$backtopage.'">';
-	print '<input type="hidden" name="id" value="'.$object->id.'">';
+$form = new Form($db);
+$iasobject=new iaskpis($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php [hunk 14]
```diff
@@ -110,715 +99,458 @@
...
+$iasstatic=new iaskpis($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_assign.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_card.php [hunk 15]
```diff
@@ -96,6 +81,13 @@
 include DOL_DOCUMENT_ROOT.'/core/actions_fetchobject.inc.php';  // Must be include, not include_once  // Must be include, not include_once. Include fetch and fetch_thirdparty but not fetch_optionals
 
 
+/*rehab.fawzy dev.php1*/
+/*start code //Get parameters*/
+$iasobject = new iaskpis($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_card.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_card.php [hunk 16]
```diff
@@ -110,110 +102,201 @@
...
+//
+//                        $object->$key=GETPOST($key,'alpha');
+//                        if ($val['notnull'] && $object->$key == '')
+//                        {
+//                            $error++;
+//                            setEventMessages($langs->trans("ErrorFieldRequired",$langs->transnoentitiesnoconv($val['label'])), null, 'errors');
+//                        }
+//                    }
+
+                    $iasobject = new iaskpis($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_card.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_card.php [hunk 17]
```diff
@@ -110,110 +102,201 @@
...
+//              {
+//                    $object->$key=GETPOST($key,'alpha');
+//                    if (in_array($key, array('rowid', 'entity', 'date_creation', 'tms', 'import_key'))) continue;
+//                    if ($val['notnull'] && $object->$key == '')
+//                    {
+//                        $error++;
+//                        setEventMessages($langs->trans("ErrorFieldRequired",$langs->transnoentitiesnoconv($val['label'])), null, 'errors');
+//                    }
+//              }
+                $iasobject = new iaskpis($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_card.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_card.php [hunk 18]
```diff
@@ -110,110 +102,201 @@
...
+                }
+        }
+
+        // Si Validation de la demande
+        if ($action == 'confirm_draft' && GETPOST('confirm') == 'yes')
         {
-            $object->$key=GETPOST($key,'alpha');
-            if (in_array($key, array('rowid', 'entity', 'date_creation', 'tms', 'import_key'))) continue;
-            if ($val['notnull'] && $object->$key == '')
+            $iasobject = new iaskpis($db);
```
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/kpi/kpi_card.php]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql (hunk 4 above)
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: done but change : 
desc=kpi_desc
name = kpi_name
status CHAR (1) = status INTEGER (1)

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: not work

Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: done changes all name

Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: not work

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql (hunk 2 above)
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: done create the same table that you sent me

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_iaskpis_assign.sql (hunk 3 above)
Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_iaskpis_assign.sql]: done

File: portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql (hunk 1 above)
Comment (hythamsoliman81) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: I will keep this comment till we find a way to use the prefix

Comment (php-1) [reply portfolio/ifintech/v1/ierp/htdocs/custom/ihcm/sql/kpi/llx_ias_kpi.sql]: ok

**Pull request merged**